cgen error when using generic interface

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
```
OS: macos, macOS, 11.6, 20G165
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.0.0 (clang-1300.0.29.3)

getwd: /Users/nedpals/Documents/coding
vmodules: /Users/nedpals/.vmodules
vroot: /Users/nedpals/Downloads/v
vexe: /Users/nedpals/Downloads/v/v
vexe mtime: 2022-06-06 05:15:13
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 ce77187.df80b33

Git version: git version 2.33.0
Git vroot status: weekly.2022.22-53-gdf80b33d (1 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-macos-arm64 173c526e
```

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```
module main

interface TypeFactory<T> {
   get_type(type_name string) T
}

enum NodeType {
   unknown
   expression
   statement
   literal
}

struct EnumTypeFactory {}
fn (f EnumTypeFactory) get_type(type_name string) NodeType {
   return match type_name {
      'expression' { NodeType.expression }
      'statement' { NodeType.statement }
      'literal' { NodeType.literal }
      else { NodeType.unknown }
   }
}

struct Node<T> {
   factory TypeFactory<T>
   type_name NodeType
}

fn new_node<T>(type_name string, factory TypeFactory<T>) Node<T> {
   return Node<T>{
     factory: factory
     type_name: factory.get_type(type_name)
   }
}

fn main() {
   root := new_node<NodeType>('literal', EnumTypeFactory{})
   println(root)
   println(new_node<NodeType>('expression', root.factory))
}
```

**What did you expect to see?**
```
Node<NodeType>{
    factory: EnumTypeFactory{}
    type_name: literal
}

Node<NodeType>{
    factory: EnumTypeFactory{}
    type_name: expression
}
```
Or similar to this
**What did you see instead?**
```
/tmp/v_501/test.1235971688113178873.tmp.c:2001:9: error: implicit declaration of function 'main__EnumTypeFactory_get_type_T_main__NodeType' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        return main__EnumTypeFactory_get_type_T_main__NodeType(*f, type_name);
               ^
/tmp/v_501/test.1235971688113178873.tmp.c:2196:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
1 warning and 1 error generated.
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5742,7 +5742,7 @@ static inline __shared__$interface_name ${shared_fn_name}(__shared__$cctype* x) 
 
 			for method in methods {
 				mut name := method.name
-				if inter_info.parent_type.has_flag(.generic) {
+				if method.generic_names.len > 0 && inter_info.parent_type.has_flag(.generic) {
 					parent_sym := g.table.sym(inter_info.parent_type)
 					match parent_sym.info {
 						ast.Struct, ast.Interface, ast.SumType {
@@ -5759,7 +5759,7 @@ static inline __shared__$interface_name ${shared_fn_name}(__shared__$cctype* x) 
 				}
 				// .speak = Cat_speak
 				if st_sym.info is ast.Struct {
-					if st_sym.info.parent_type.has_flag(.generic) {
+					if method.generic_names.len > 0 && st_sym.info.parent_type.has_flag(.generic) {
 						name = g.generic_fn_name(st_sym.info.concrete_types, method.name,
 							false)
 					}

--- a/vlib/v/tests/generics_interface_with_non_generic_method_test.v
+++ b/vlib/v/tests/generics_interface_with_non_generic_method_test.v
@@ -1,0 +1,45 @@
+module main
+
+interface TypeFactory<T> {
+	get_type(type_name string) T
+}
+
+enum NodeType {
+	unknown
+	expression
+	statement
+	literal
+}
+
+struct EnumTypeFactory {}
+
+fn (f EnumTypeFactory) get_type(type_name string) NodeType {
+	return match type_name {
+		'expression' { NodeType.expression }
+		'statement' { NodeType.statement }
+		'literal' { NodeType.literal }
+		else { NodeType.unknown }
+	}
+}
+
+struct Node<T> {
+	factory   TypeFactory<T>
+	type_name NodeType
+}
+
+fn new_node<T>(type_name string, factory TypeFactory<T>) Node<T> {
+	return Node<T>{
+		factory: factory
+		type_name: factory.get_type(type_name)
+	}
+}
+
+fn test_generic_interface_with_non_generic_method() {
+	root1 := new_node<NodeType>('literal', EnumTypeFactory{})
+	println(root1)
+	assert root1.type_name == .literal
+
+	root2 := new_node<NodeType>('expression', root1.factory)
+	println(root2)
+	assert root2.type_name == .expression
+}

